OsmAnd/src/net/osmand/plus/myplaces/tracks/filters/viewholders/FilterFolderViewHolder.kt

```diff
@@ -147,96 +184,187 @@ class FilterFolderViewHolder(itemView: View, nightMode: Boolean) :
+						updateAllFoldersCheck()
+						holder.checkBox.isChecked = filter.isFolderSelected(folderName)
+						this.notifyItemChanged(correctedPosition)
+						this.notifyItemChanged(0)
+						filterChangedListener?.onFilterChanged()
 					}
+					holder.checkBox.isChecked = filter.isFolderSelected(folderName)
+					var itemsCount = filter.allFoldersCollection[folderName] ?: 0
+					holder.count.text = itemsCount.toString()
+					holder.itemView.isEnabled = itemsCount > 0
+					holder.checkBox.isEnabled = itemsCount > 0
+					holder.checkBox.setOnCheckedChangeListener(null)
 					holder.icon.setImageDrawable(
 						app.uiUtilities.getPaintedIcon(
 							R.drawable.ic_action_folder,
 							app.getColor(R.color.icon_color_default_light)))
 				}
 
 				is AllFoldersViewHolder -> {
-					filter?.let {
-						holder.switch.isSelected = it.isAllFoldersSelected
-						if (holder.switch.isSelected) {
-							holder.icon.setImageDrawable(
-								app.uiUtilities.getActiveIcon(
-									R.drawable.ic_action_folder,
-									nightMode))
-						} else {
-							holder.icon.setImageDrawable(
-								app.uiUtilities.getPaintedIcon(
-									R.drawable.ic_action_folder,
-									app.getColor(R.color.icon_color_default_light)))
-						}
+					isAllFoldersBeingSet = true
+					holder.switch.state = if (filter.isAllFoldersSelected ||
+						filter.selectedFolders.size == items.size) {
+						holder.switch.isChecked = true
+						ThreeStateCheckbox.State.CHECKED
+					} else if (filter.selectedFolders.size > 0) {
+						holder.switch.isChecked = false
+						ThreeStateCheckbox.State.MISC
+					} else {
+						holder.switch.isChecked = false
+						ThreeStateCheckbox.State.UNCHECKED
+					}
+					isAllFoldersBeingSet = false
+					if (holder.switch.state != ThreeStateCheckbox.State.UNCHECKED) {
+						holder.icon.setImageDrawable(
+							app.uiUtilities.getActiveIcon(
+								R.drawable.ic_action_folder,
+								nightMode))
+					} else {
+						holder.icon.setImageDrawable(
+							app.uiUtilities.getPaintedIcon(
+								R.drawable.ic_action_folder,
+								app.getColor(R.color.icon_color_default_light)))
 					}
 					holder.switch.setOnCheckedChangeListener { _, isChecked ->
-						onAllFolderSelected(isChecked)
+						onAllFolderSelected(holder.switch, isChecked)
 					}
 				}
 
 				is ShowAllViewHolder -> {
-					holder.title.setText(if (fullFoldersListExpanded) R.string.shared_string_show_less else R.string.shared_string_show_all)
+					holder.title.setText(R.string.shared_string_show_all)
 					holder.itemView.setOnClickListener {
-						fullFoldersListExpanded = !fullFoldersListExpanded
-						notifyDataSetChanged()
+						val folderFilter = TrackFolderFilter(null)
+						folderFilter.initWithValue(filter)
+						folderFilter.setFullFoldersCollection(filter.allFoldersCollection)
+						val folderAdapter = FolderAdapter(app, nightMode, null, null)
+						folderAdapter.filter = folderFilter
+						folderAdapter.isExpanded = true
```

Comment:
Looks strange, folderAdapter.isExpanded set to true for showing all elements from this adapter in another screen